docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,322 @@
+#### Prometheus-metrics-adapter precludes the use of sources of custom metrics
+With current K8s design, only one API extension can serve custom metrics. Per this GEP, the custom metrics extension
+point is occupied by prometheus-metrics-adapter.
+
+#### Heavyweight Prometheus infrastructure feeds signal back into cluster control loop
+The proposed signal flow brings two Prometheus instances (shoot prometheus and seed aggregate prometheus) into the shoot
+cluster control loop. This has major implications on reliability. The alternative is a small, custom-built component
+registered as seed custom metrics API extension, which directly calls the ShootKapi metrics endpoints and outputs
+request rate.
+
+The current approach was chosen solely because it leverages an existing open source component and avoids authoring a
+custom one. The historical record of Gardener Prometheus components reliability is deemed acceptable for the purpose.  
+
+The proposed complex signal path introduces additional latency to scaling response. At present, the resulting scaling
+latency is deemed satisfactory.
+
+While the scope of this GEP is limited to using the Prometheus signal path, the proposed design allows for an easy
+substitution with a signal path based on aforementioned simple custom component alternative. This supports future
+changes based on accumulated field experience, or evolving solution needs. 
+
+#### Observed fluctuations in actual compute efficiency affecting ShootKapi workloads
+There is a known source of horizontal scaling inefficiency, which presents an opportunity for substantial further
+improvement. It is not addressed by this proposal and instead deferred to a separate implementation stage, as it
+requires a strictly incremental effort on top of this GEP.
+
+An approximate 50% reduction (factor 0.5) of the effective compute power per unit of CPU was occasionally observed on
+some productive ShootKapi instances. The change occurs with a quick cutoff transition, and has been observed for periods
+or up to 2 hours. This variance currently forces substantial safety margin upon HPA's target metric value, effectively
+necessitating more replicas to be deployed than required during usual operation.
+
+The source is believed to be competition for compute resources with sibling workloads, either on the K8s, or
+infrastructure level, but the circumstances have not been conductive to investigation after the fact.
+A subsequent stage, outside the scope of this GEP is planned to focus on understanding the root cause, and further
+optimizing HPA's target metric value accordingly.
+
+Once this biggest source of inefficiency has been resolved, yet another round of scaling performance fine-tuning
+is planned, where instead of overall request rate, HPA will be driven by a weighted average, reflecting the different
+compute cost of different request categories (e.g. a cluster-scoped LIST is more cpu intensive than a resource-scoped PUT)
```

Comment:
Couldn't the different resource consumptions of request categories be another reason for a deadlock situation? 
Assuming there is a cluster which runs many resource intensive cluster-scoped list events with label selectors on tens of thousands config maps. In this case the maximum size of the ShootKapi instances (node capacity) could be reached, before the request rate is high enough to trigger HPA scaling up horizontally.
On the other side, if the request limit is too low and there are many "cheap" requests, HPA would scale up too early so that there are too many small ShootKapi instances which are not running in a resource efficient way.

Reply:
This precisely is the motivation behind moving to a weighted average in the subsequent optimisation stage.

With HPA driven by a signal which does not predict resource demand with perfect accuracy, there is a tradeoff between:
- minimising the overhead which is the horizontal safety margin of extra requested replicas (efficiency)
- risk of providing insufficient number of replicas under atypical load profiles (reliability)

In the end, I expect it will come down to simple cost-benefit optimisation. We're starting with conservative 
(low) requests/replica, and then field data will show how far we can optimise.

Reply:
Closing the conversation to keep the PR manageable. Please don't hesitate to reopen if anything else comes to mind.